tox.ini

```diff
@@ -13,6 +13,12 @@ commands = pytest -v -ra --cov=gluetool --cov-report=html:coverage-report {posar
 [testenv:py27-static-analysis]
 commands = pytest -v -ra --pylint --flake8 -k 'not test_' {posargs}
 
+[testenv:py36-unit-tests]
+commands = pytest -v -ra --ignore=docs/source/module_parsers.py {posargs}
+
+[testenv:py36-static-analysis]
+commands = pytest -v -ra --pylint --flake8 -k 'not test_' --ignore=docs/source/module_parsers.py {posargs}
+
```

Comment:
It might be possible to merge `py27-*` with their `py36-*` counterparts, however I don't know how. But tox hase some abilities, e.g. `envlist = {py27,py36}-{unit-tests,static-analysis}`. I'll find the answer.

Reply:
Here you go - condensed, and with one exception I really like it. When you start with more than one Python version, tox starts to prove very useful:

``` ini
[tox]
envlist = {py27,py36}-{static-analysis,unit-tests,codecov}

[testenv]
# Tox should set correct Python version given the environment. let's trust it knows how.
#basepython =
#  py27: python2.7
#  py36: python3.6

# Use a single env directory for all environments but separate by Python versions.
#
# It could be easier to just say:
#
#   envdir = {toxworkdir}/{basepython}
#
# but, since it's hardcoded in Tox, basepython is not set when envdir is being examined,
# therefore, the replacement fails, hence the explicit names.
envdir =
  py27: {toxinidir}/.tox/py27
  py36: {toxinidir}/.tox/py36

# Don't spoil our nice virtualenvs with system packages
sitepackages = False

# Just test dependencies - gluetool will pull in its own dependencies when being installed
deps = -rtest-requirements.txt

# Capture coverage per Python version
setenv = COVERAGE_FILE={envdir}/.coverage

# Pass necessary env vars to let CI and coverage gathering play together nicely
passenv = CI TRAVIS TRAVIS_* CODECOV_TOKEN

# Conditionals don't support multiline branches, therefore codecov upload is a separate step.
# but that's fine, thanks to options we give it it knows where to find coverage (specific for
# Python version), and we even tell it to add Python version as a flag - keeping coverage
# separate even in Codecov seems to be a good starting point, we can always start merging
# them later.
#
# And there's an advantage: the codecov upload does not run when one runs *-unit-tests env locally.
# It'd fail anyway, since one does not usually have CODECOV_TOKEN set, and it's still possible
# to run it explicitly to upload one's local coverage in case of need.
commands =
  static-analysis: pytest -v -ra --pylint --flake8 -k 'not test_' {posargs}
  unit-tests: pytest -v -ra --cov=gluetool --cov-report=html:{envdir}/coverage-report {posargs}
  codecov: codecov --file={envdir}/.coverage --flags={basepython}

# [testenv:py27-doctest]
# whitelist_externals = ansible-playbook
# commands = /usr/bin/ansible-playbook generate-docs.yml -e generate_dir="{envtmpdir}/docs"

# [testenv:py27-static-analysis-coala]
# whitelist_externals = docker
# commands = docker run -ti --rm -v {toxinidir}:/gluetool:z --workdir=/gluetool coala/base coala -c /gluetool/.coafile --non-interactive
```

Works like a charm for me, just run `tox` to run all environments. It will run `codecov` upload as well, which would fail as you don't have CODECOV_TOKEN env var set locally, but `tox` will report it as a success - which is perfectly fine, it works perfectly within Travis.

Reply:
Thanks @happz, incorporated